### PR TITLE
qgis: Enable mapserver

### PR DESCRIPTION
**Pull request by bgamari**
## Description of changes

Here we enable QGis's internal [tileserver] implementation. I have found this useful and the size of the binary itself is negligible (around 300kB).

[tileserver]: https://docs.qgis.org/3.28/en/docs/server_manual

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by imincik — CHANGES_REQUESTED** [review]
@bgamari , thank you very much for contribution. I am interested in server build too, but it would be much better if we build separate server package. Server package should be built for LTR version as well and nixos test is should be included with PR.

I am happy to help you if you are interested in enhancing your PR.

**Comment by bgamari** [comment]
> @bgamari , thank you very much for contribution. I am interested in server build too, but it would be much better if we build separate server package. Server package should be built for LTR version as well and nixos test is should be included with PR.
> 
> I am happy to help you if you are interested in enhancing your PR.

@imincik do you mean a separate package or a separate output?

**Comment by imincik** [comment]
> do you mean a separate package or a separate output?

That's a good question, I am not sure. Maybe separate `server` output is good enough. What do you think ?

File: pkgs/applications/gis/qgis/unwrapped.nix [hunk 1]
```diff
@@ -150,6 +150,7 @@ in mkDerivation rec {
   '';
 
   cmakeFlags = [
+    "-DWITH_SERVER=True"
     "-DWITH_3D=True"
     "-DWITH_PDAL=TRUE"
     "-DENABLE_TESTS=False"
```
Comment (sikmir) [inline pkgs/applications/gis/qgis/unwrapped.nix:156]:
```suggestion
    (lib.cmakeBool "WITH_SERVER" true)
    (lib.cmakeBool "WITH_3D" true)
    (lib.cmakeBool "WITH_PDAL" true)
    (lib.cmakeBool "ENABLE_TESTS" false)
```

**Comment by bgamari** [comment]
> That's a good question, I am not sure. Maybe separate server output is good enough. What do you think ?

I suspect that a separate output is really the only viable option. From a cursory inspection of QGIS's build system, it doesn't appear to be possible to build the core library and then later (in a separate tree) link the desktop application/server against that library. Consequently, I think the only option is to build all three and then carefully separate them into separate outputs.

However, I'm not sure I feel confident enough in my knowledge of QGIS's internal structure to be able to safely do this.

**Comment by bgamari** [comment]
I have looked more closely at the output of the derivation and have concluded that I don't feel comfortable performing an output split. There are simply too many files which I, as a relative casual user of QGIS, cannot easily determine the usage of (e.g. should `share/qgis/svg` belong in the `lib` or `bin` output? the same for `lib/qgis/plugins/libplugin_geometrychecker.so`? `lib/qgis/qgiscrashhandler`?) 

I think either we should merge this as-is or we decide not to merge at all. The risk (and maintainability challenge) of performing an output split strikes me as not justified by the very minor size increase that enabling the server component.

I have, however, integrated @sikmir's suggestion and propagated the change to the LTR derivation.

**Comment by imincik** [comment]
What about following [QGIS Debain](https://github.com/qgis/QGIS/tree/master/debian) packaging and put everything what they in `*-server.install` files in to our server output ?

**Comment by bgamari** [comment]
> What about following [QGIS Debain](https://github.com/qgis/QGIS/tree/master/debian) packaging and put everything what they in `*-server.install` files in to our server output ?

I suppose you mean `qgis-server-*.install`? While I'm not necessarily opposed, I do really question what problem we are solving by doing so. It seems like we are making maintenance work for ourselves for fairly little tangible benefit.

**Comment by imincik** [comment]
> I do really question what problem we are solving by doing so. It seems like we are making maintenance work for ourselves for fairly little tangible benefit.

I think the main reason is to keep QGIS package size down. I guess that vast majority of QGIS package users are using only desktop interface. Also, I think the list of files belonging to server is not changing very often so the maintenance cost wouldn't be very high.

**Comment by imincik** [comment]
> > What about following [QGIS Debain](https://github.com/qgis/QGIS/tree/master/debian) packaging and put everything what they in `*-server.install` files in to our server output ?
> 
> I suppose you mean `qgis-server-*.install`? While I'm not necessarily opposed, I do really question what problem we are solving by doing so. It seems like we are making maintenance work for ourselves for fairly little tangible benefit.

Yes, sorry. I fixed that in my comment.

**Comment by imincik** [comment]
@bgamari , I am experimenting with QGIS server build in https://github.com/NixOS/nixpkgs/pull/267301 .

**Comment by imincik** [comment]
This PR is now replaced by https://github.com/NixOS/nixpkgs/pull/267301

**Pull request closed without merging**